### PR TITLE
#15-add-collection-pagination

### LINKED ISSUE #15: Add pagination to plugin collections.
We need to add in pagination arguments and functionality to the plugins endpoint in `get_items()`.

### PR DESCRIPTION
**Pull request by BE-Webdesign**
Fixes #15.  This is a first stab at pagination for the collection.
Because we can't make use of WP_Query for plugins, we do manual
pagination utilizing `array_splice()`.

### CODE REVIEW

File: lib/class-wp-rest-plugins-controller.php [hunk 1]
```diff
@@ -57,16 +57,71 @@ public function get_items( $request ) {
 		$data = array();
 
 		require_once ABSPATH . '/wp-admin/includes/plugin.php';
-		foreach ( get_plugins() as $obj ) {
+		$plugins = get_plugins();
+
+		// Exit early if empty set.
+		if ( empty( $plugins ) ) {
+			return rest_ensure_response( $data );
```
Comment (joehoyle) [inline lib/class-wp-rest-plugins-controller.php]: let's return `return array()` here, simpler to read - no need to `rest_ensure_response`

File: lib/class-wp-rest-plugins-controller.php [hunk 2]
```diff
@@ -194,7 +249,16 @@ public function get_item_schema() {
 	}
 
 	public function get_collection_params() {
-		return array();
+		$params = parent::get_collection_params();
+
+		$params['offset'] = array(
```
Comment (joehoyle) [inline lib/class-wp-rest-plugins-controller.php]: Can we use `page` / `per_page` like we do with the other endpoints here?

**Comment by joehoyle** [comment]
@BE-Webdesign great start! I'm wondering if we _need_ pagination here or not. Given that `get_plugins()` returns all data, we aren't saving anything on the DB etc by adding pagination. It does make it more consistent with other collection endpoints, however this could be a case where we don't need to.

File: lib/class-wp-rest-plugins-controller.php (hunk 2 above)
Comment (BE-Webdesign) [inline lib/class-wp-rest-plugins-controller.php]: They are being "inherited" (not really but kind of) from `WP_REST_Controller` via `$params = parent::get_collection_params()`.  My interpretation of the API led me to believe that is how these pagination arguments were meant to be brought in.  Should we explicitly add them here or maybe add a comment above the call to the controller class explaining that `page / per_page` come from there?

Comment (joehoyle) [inline lib/class-wp-rest-plugins-controller.php]: Oh, as `page` and `per_page` are already brought in from `parent::get_collection_params()` then that's fine. As for `offset`, I hadn't realized we also support that in the other controllers too. Who knows what happens if you specify `offset` with `page`!

**Comment by joehoyle** [comment]
I think we can roll with this, thanks @BE-Webdesign!

**Comment by BE-Webdesign** [comment]
>  I'm wondering if we need pagination here or not.

I think we should still support pagination, not only for consistency.  It still is a minor performance enhancement as you would only run `prepare_item_for_response`.  In my tests it actually adds up to hundreds on ms on response times when you have hundreds of plugins vs only ten being returned in the collection.

**Comment by joehoyle** [comment]
Makes sense, consistency FTW!

**Pull request merged**